Create custom DiagnosticsNode type for DevTools deep links.

kenzieschmoll opened the pull request:
IDEs will look for this specific diagnostic node type in order to prompt the user to inspect the errors in DevTools.
cc @helin24 @DanTup

kenzieschmoll committed:
Create custom DiagnosticsNode type for DevTools deep links.

kenzieschmoll committed:
add test

jacob314 commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1801,6 +1801,19 @@ class MessageProperty extends DiagnosticsProperty<void> {
+  /// [value] will return the String representation of the DevTools url.
+  ///
+  /// The [description] and [value] arguments must not be null.
+  DevToolsDeepLinkDiagnosticsNode(String description, String value) :
```
Consider renaming parameter `value` to `url` and making it named.

jacob314 commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1801,6 +1801,19 @@ class MessageProperty extends DiagnosticsProperty<void> {
        super(name, null, description: message, style: style, level: level);
 }
 
+/// Debugging message for DevTools deep links.
```
Top level comments should ideally say something more than just the words from the class name in another order. Clarify why it makes sense to indicate that a link is to Flutter DevTools vs any other deep link you might have in the app.
Change the class name to be consistent with the other class names.
For example
DevToolsDeepLinkProperty would be better.

jacob314 replied:
Move this class from diagnostics.dart to widget_inspector.dart. This isn't general purpose enough to be in foundation

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -2934,8 +2934,9 @@ Iterable<DiagnosticsNode> _describeRelevantUserCode(Element element) {
           WidgetInspectorService.instance._devToolsInspectorUriForElement(target);
       if (devToolsInspectorUri != null) {
-        devToolsDiagnostic = DiagnosticsNode.message(
+        devToolsDiagnostic = DevToolsDeepLinkDiagnosticsNode(
```
consider moving some of the logic from _devToolsInspectorUriForElement into DevToolsDeepLink as it is tightly coupled and that would make the DevToolsDeepLink class a little less surprising than if its public constructor takes an arbitrary URL which might or might not be from DevTools.

goderbauer commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1801,6 +1801,19 @@ class MessageProperty extends DiagnosticsProperty<void> {
+class DevToolsDeepLinkDiagnosticsNode extends StringProperty {
+  /// Create a diagnostics property that displays a deep link to DevTools.
+  ///
+  /// [value] will return the String representation of the DevTools url.
```
nit: doc sentences should be grammatically correct and start with a capital letter, e.g. "The [value] will..."

kenzieschmoll committed:
Review comments.

kenzieschmoll commented:
Addressed review comments and refactored some things. Represented the diagnostic value as a map so I can add useful parameters for IDEs to consume (other wise they would have had to parse information from the url). PTAL.

kenzieschmoll committed:
fix test

kenzieschmoll committed:
revert change

kenzieschmoll committed:
improve test.

jacob314 commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -2958,6 +2940,115 @@ Iterable<DiagnosticsNode> _describeRelevantUserCode(Element element) {
+/// diagnostic pertains to.
+/// `objectId` is an optional id specifying the data object in Flutter DevTools
+/// that this diagnostic pertains to.
+class DevToolsDeepLinkProperty extends DiagnosticsProperty<Map<String, String>> {
```
This should not be a property on Map<String, String>
instead make 
DevToolsDeepLink extend Diagnosticable.
Then it is clean enough to add the extra json you want and specify the tostring you needed.

kenzieschmoll committed:
Simplify

jacob314 commented on packages/flutter/test/widgets/widget_inspector_test.dart:
```diff
@@ -2835,6 +2835,31 @@ class _TestWidgetInspectorService extends TestWidgetInspectorService {
+          'missingIfNull': false,
+          'propertyType': 'String',
+          'defaultLevel': 'info',
+          'value': 'the-deep-link',
```
nit: make this
'http://the-deeplink/'

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)

goderbauer commented on packages/flutter/lib/src/widgets/widget_inspector.dart:
```diff
@@ -2958,6 +2959,27 @@ Iterable<DiagnosticsNode> _describeRelevantUserCode(Element element) {
   return nodes;
 }
 
+
```
nit: just one blank line

goderbauer reviewed (approved):
LGTM
